logstash-core/lib/logstash/output_delegator.rb

```diff
@@ -1,192 +1,60 @@
-# encoding: utf-8
-require "concurrent/atomic/atomic_fixnum"
-java_import "java.util.concurrent.CopyOnWriteArrayList"
+require "logstash/output_delegator_strategy_registry"
+
+require "logstash/output_delegator_strategies/shared"
+require "logstash/output_delegator_strategies/single"
 
-# This class goes hand in hand with the pipeline to provide a pool of
-# free workers to be used by pipeline worker threads. The pool is
-# internally represented with a SizedQueue set the the size of the number
-# of 'workers' the output plugin is configured with.
-#
-# This plugin also records some basic statistics
 module LogStash class OutputDelegator
-  attr_reader :workers, :config, :threadsafe
+  attr_reader :metric, :metric_events, :strategy, :namespaced_metric, :metric_events , :plugin_args
 
-  # The *args this takes are the same format that a Outputs::Base takes. A list of hashes with parameters in them
-  # Internally these just get merged together into a single hash
-  def initialize(logger, klass, default_worker_count, metric, *plugin_args)
+  def initialize(logger, output_class, metric, xopts={}, plugin_args={})
     @logger = logger
-    @threadsafe = klass.threadsafe?
-    @config = plugin_args.reduce({}, :merge)
-    @klass = klass
-    @workers = java.util.concurrent.CopyOnWriteArrayList.new
-    @default_worker_count = default_worker_count
-    @registered = false
-
-    # Create an instance of the input so we can fetch the identifier
-    output = @klass.new(@config)
-
-    # Scope the metrics to the plugin
-    namespaced_metric = metric.namespace(output.plugin_unique_name.to_sym)
-    output.metric = namespaced_metric
+    @output_class = output_class
+    @metric = metric
+    @xopts = xopts # Options specific to the given class
+    @plugin_args = plugin_args
+    build_strategy!
 
-    @metric_events = namespaced_metric.namespace(:events)
-    namespaced_metric.gauge(:name, config_name)
-
-    @events_received = Concurrent::AtomicFixnum.new(0)
+    @id = plugin_args["id"]
+    @namespaced_metric = metric.namespace("#{@output_class.config_name}_#{@id}".to_sym)
```

Comment:
Consider using the method `config_name` for the metric namespace, further more that would remove the need to create an ivar on line 17.